### PR TITLE
Env

### PR DESCRIPTION
**Pull request by Kimchelly**
- Added basic checks for semant on declare, assign, binop, and unop
- Added code generation for on declare, assign, binop, and unop
- Note: I'm pretty sure something like `int x = y = 3;` won't evaluate correctly, so this will need fixing.
- Changed spaces to tabs

### CODE REVIEW

File: codegen.ml [hunk 1]
```diff
@@ -38,124 +39,256 @@ let str_t = L.pointer_type i8_t;;
...
+		with Not_found ->
+		try
+			Hash.find local_params id
+		with Not_found ->
+			raise(Failure("Unknown variable " ^ id))
 
 and func_lookup fname = match (L.lookup_function fname codegen_module) with
-      None -> raise (Failure(" function " ^ fname ^ " does not exit."))
-    | Some func -> func 
+	  None -> raise (Failure(" function " ^ fname ^ " does not exit."))
```
Comment (kycwuwu) [inline codegen.ml]: I'm sorry this was from me, but the function is supposed to exist not exit OOPS

File: codegen.ml [hunk 2]
```diff
@@ -38,124 +39,256 @@ let str_t = L.pointer_type i8_t;;
...
-    | SStringLit(s) -> string_gen llbuilder s
-    | SCall("print", sexpr_list) -> call_gen llbuilder "printf" sexpr_list
-    A.Void
-    | SNoexpr -> L.const_int i32_t 0
-    | _ -> raise (Failure ("Expression type not recognized.")) 
+	  SIntLit(i) -> L.const_int i32_t i
+	| SBoolLit(b) -> if b then L.const_int i1_t 1 else L.const_int i1_t 0
+	| SFloatLit(f) -> L.const_float f_t f
+	| SStringLit(s) -> string_gen llbuilder s
+	| SId(id, typ) -> print_string id; id_gen llbuilder id true
```
Comment (kycwuwu) [inline codegen.ml]: Lone print_string

File: codegen.ml [hunk 3]
```diff
@@ -38,124 +39,256 @@ let str_t = L.pointer_type i8_t;;
...
-        "printf" llbuilder
-    
+	let params = List.map (fun expr -> sexpr_gen llbuilder expr) sexpr_list
+	in
+	L.build_call (func_lookup "printf") 
+		(Array.of_list ((sexpr_gen llbuilder (SStringLit("%s")))::params))
+		"printf" llbuilder
+
+(* Generates a local variable declaration *)
+and local_var_gen llbuilder typ id =
```
Comment (kycwuwu) [inline codegen.ml:219]: do we even need to define noop in here? where is it ever used?
As far as I can see, we only need to get ltyp from the two cases to use in making alloc block

File: semant.ml [hunk 4]
```diff
@@ -6,257 +6,493 @@ open Sast
...
+	let sel, env = get_sexprl env el in
+	let check_param formal param =
+		let f_typ = match formal with Formal(t, _) -> t | _ -> Void in
+		let p_typ = get_type_from_sexpr param in
+		if (f_typ = p_typ) then param
+		else
+			raise(Failure("Incorrect type passed to function")) 
+			in
+	let handle_params (formals) params =
+		(* Should do some formal/param pattern matching later *)
```
Comment (kycwuwu) [inline semant.ml]: calling check_param handles the type checking :)

**Review by kycwuwu — CHANGES_REQUESTED** [review]

File: semant.ml (hunk 4 above)
Comment (Kimchelly) [reply semant.ml]: I didn't write this

File: codegen.ml (hunk 3 above)
Comment (Kimchelly) [reply codegen.ml:219]: Fixed

File: codegen.ml (hunk 1 above)
Comment (Kimchelly) [reply codegen.ml]: Fixed

File: codegen.ml (hunk 2 above)
Comment (Kimchelly) [reply codegen.ml]: Well it was so lonely I killed it

**Pull request closed without merging**